base/dsfs/commit.go

```diff
@@ -25,86 +24,64 @@ var Timestamp = func() time.Time {
-	BodyDefault = BodyAction("default")
+	BodyDefault BodyAction = "default"
 	// BodySame means that the bodies are the same, no need to compare
-	BodySame = BodyAction("same")
+	BodySame BodyAction = "same"
 	// BodyTooBig means the body is too big to directly compare, and should use
 	// some other method
-	BodyTooBig = BodyAction("too_big")
+	BodyTooBig BodyAction = "too_big"
 )
 
-// DerefCommit derferences a dataset's Commit element if required
-// should be a no-op if ds.Structure is nil or isn't a reference
-func DerefCommit(ctx context.Context, store qfs.Filesystem, ds *dataset.Dataset) error {
-	if ds.Commit != nil && ds.Commit.IsEmpty() && ds.Commit.Path != "" {
-		cm, err := loadCommit(ctx, store, ds.Commit.Path)
-		if err != nil {
-			log.Debug(err.Error())
-			return fmt.Errorf("loading dataset commit: %w", err)
+func commitFileAddFunc(ctx context.Context, privKey crypto.PrivKey, publisher event.Publisher) writeComponentFunc {
+	return func(src qfs.Filesystem, dst qfs.MerkleDagStore, prev, ds *dataset.Dataset, added qfs.Links, sw *SaveSwitches) error {
+		if ds.Commit == nil {
+			return errNoComponent
 		}
-		cm.Path = ds.Commit.Path
-		ds.Commit = cm
-	}
-	return nil
-}
 
-// loadCommit assumes the provided path is valid
-func loadCommit(ctx context.Context, fs qfs.Filesystem, path string) (st *dataset.Commit, err error) {
-	data, err := fileBytes(fs.Get(ctx, path))
-	if err != nil {
-		log.Debug(err.Error())
-		return nil, fmt.Errorf("loading commit file: %s", err.Error())
-	}
-	return dataset.UnmarshalCommit(data)
-}
-
-func commitFileAddFunc(privKey crypto.PrivKey, pub event.Publisher) addWriteFileFunc {
-	return func(ds *dataset.Dataset, wfs *writeFiles) error {
-		if ds.Commit == nil {
-			return nil
+		if evtErr := publisher.Publish(ctx, event.ETDatasetSaveProgress, event.DsSaveEvent{
+			Username:   ds.Peername,
+			Name:       ds.Name,
+			Message:    "finalizing",
+			Completion: 0.9,
+		}); evtErr != nil {
+			log.Debugw("publish event errored", "error", evtErr)
 		}
 
-		hook := func(ctx context.Context, f qfs.File, added map[string]string) (io.Reader, error) {
-			if evtErr := pub.Publish(ctx, event.ETDatasetSaveProgress, event.DsSaveEvent{
-				Username:   ds.Peername,
-				Name:       ds.Name,
-				Message:    "finalizing",
-				Completion: 0.9,
-			}); evtErr != nil {
-				log.Debugw("publish event errored", "error", evtErr)
-			}
+		log.Debugw("writing commit file", "bodyAction", sw.bodyAct, "force", sw.ForceIfNoChanges, "fileHint", sw.FileHint)
 
-			if cff, ok := wfs.body.(*computeFieldsFile); ok {
-				updateScriptPaths(ds, added)
+		updateScriptPaths(dst, ds, added)
 
-				if err := confirmByteChangesExist(cff.ds, cff.prev, added, wfs.body.FullPath(), cff.sw.ForceIfNoChanges); err != nil {
-					return nil, fmt.Errorf("error saving: %w", err)
-				}
+		if err := confirmByteChangesExist(ds, prev, sw.ForceIfNoChanges, dst, added); err != nil {
+			return fmt.Errorf("saving: %w", err)
```

Comment:
If this error is seen by a user, it reads as ambiguous whether a save actually happened. `saving` is meant to be read like a category with "no changes" as an error, but it could be seen as a verb with "no changes" as purely informational. Consider changing to `saving failed: no changes`.

Reply:
agree. I just didn't like leading with "error", added in your suggestion